In [4]:
## import
import pandas as pd
from pathlib import Path


# import streamlit as st
# import plotly.express as px
# import plotly.graph_objects as go


TARGET_DIR = 'data'
TARGET_CSV = 'data.csv'

# BASE_DIR = Path(__file__).resolve().parent
BASE_DIR = Path.cwd().resolve().parent
DATA_PATH = BASE_DIR / TARGET_DIR / TARGET_CSV

## read_csv / file: "한국해양진흥공사_부산항_노선별_항만물류지표_KPLI.csv"
df = pd.read_csv(DATA_PATH)


df

In [5]:
df.describe()

,항만효율성(PORT_EFFICIENCY),총항해시간(TOTAL_SAILING_TIME),대기시간(WAITING_TIME),항만정시성(ON_TIME_PERFORMANCE)
count,480.000000,480.000000,480.000000,480.000000
mean,19.756458,432.292521,0.138042,6.431250
std,5.754084,415.966180,0.269674,25.602098
min,6.950000,28.970000,0.000000,-322.050000
25%,16.350000,75.985000,0.000000,1.940000
50%,18.735000,327.210000,0.030000,7.645000
75%,21.427500,624.942500,0.170000,13.950000
max,47.010000,1632.540000,3.040000,135.920000


In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 480 entries, 0 to 479
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   출발항코드(DEPARTURE_PORT_CODE)  480 non-null    str    
 1   출발항명(DEPARTURE_PORT_NAME)   480 non-null    str    
 2   도착항코드(DEST_PORT_CODE)       480 non-null    str    
 3   도착항명(DEST_PORT_NAME)        480 non-null    str    
 4   도착항국가(DEST_COUNTRY)         426 non-null    str    
 5   기준일자(DATE)                  480 non-null    str    
 6   항만효율성(PORT_EFFICIENCY)      480 non-null    float64
 7   총항해시간(TOTAL_SAILING_TIME)   480 non-null    float64
 8   대기시간(WAITING_TIME)          480 non-null    float64
 9   항만정시성(ON_TIME_PERFORMANCE)  480 non-null    float64
dtypes: float64(4), str(6)
memory usage: 37.6 KB


In [7]:
df.columns.to_list()

['출발항코드(DEPARTURE_PORT_CODE)',
 '출발항명(DEPARTURE_PORT_NAME)',
 '도착항코드(DEST_PORT_CODE)',
 '도착항명(DEST_PORT_NAME)',
 '도착항국가(DEST_COUNTRY)',
 '기준일자(DATE)',
 '항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']

In [8]:
df[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].isna().sum()

# 결측치는 존재하지않는다.

항만효율성(PORT_EFFICIENCY)        0
총항해시간(TOTAL_SAILING_TIME)     0
대기시간(WAITING_TIME)            0
항만정시성(ON_TIME_PERFORMANCE)    0
dtype: int64

In [9]:
dest_port_df = df[df['도착항코드(DEST_PORT_CODE)']!= 'KPLI']
kpli_port_df = df[df['도착항코드(DEST_PORT_CODE)']== 'KPLI']

dest_port_df['도착항코드(DEST_PORT_CODE)'].unique()
kpli_port_df['도착항코드(DEST_PORT_CODE)'].unique()
dest_port_df['기준일자(DATE)'].min(), dest_port_df['기준일자(DATE)'].max()
dest_port_df[['기준일자(DATE)','도착항코드(DEST_PORT_CODE)']].value_counts()
# kpli_port_df['기준일자(DATE)'].unique()

dest_port_df[
    (dest_port_df['기준일자(DATE)'] == '2025-10-15')
    &
    (dest_port_df['도착항코드(DEST_PORT_CODE)'] == 'AEJEA')
]
# 임의의 데이터를 추출해보니 동일일자 동일도착항의 데이터가 다름.

"""
 “월별 평균 KPI” 이 아니라
 “관측된 데이터 기준 월별 변화 추이”
"""
# df.groupby(['기준일자(DATE)', '도착항코드(DEST_PORT_CODE)']).size()

'\n “월별 평균 KPI” 이 아니라\n “관측된 데이터 기준 월별 변화 추이”\n'

In [62]:
df

df['기준일자(DATE)'].unique()

df['기준일자(DATE)'].unique()

df[df['도착항코드(DEST_PORT_CODE)']=='KPLI'].groupby(
    pd.to_datetime(df['기준일자(DATE)']).dt.to_period('M')).size()

기준일자(DATE)
2025-05    1
2025-06    5
2025-07    4
2025-08    4
2025-09    3
2025-10    6
2025-11    3
2025-12    5
2026-01    4
2026-02    4
2026-03    5
2026-04    4
2026-05    2
2026-06    4
Freq: M, dtype: int64

In [48]:
kpli_mean_dt = kpli_port_df.groupby(
    pd.to_datetime(kpli_port_df['기준일자(DATE)']).dt.to_period('M')
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

kpli_mean_dt

,항만효율성(PORT_EFFICIENCY),총항해시간(TOTAL_SAILING_TIME),대기시간(WAITING_TIME),항만정시성(ON_TIME_PERFORMANCE)
기준일자(DATE),,,,
2025-05,18.990000,239.370000,0.130000,2.290000
2025-06,20.472000,333.906000,0.086000,12.574000
2025-07,19.575000,293.200000,0.112500,3.970000
2025-08,18.922500,336.622500,0.145000,6.217500
2025-09,17.853333,308.896667,0.140000,9.540000
2025-10,18.001667,317.781667,0.145000,10.461667
2025-11,17.930000,336.493333,0.153333,7.263333
2025-12,18.690000,326.548000,0.080000,14.276000
2026-01,19.317500,334.735000,0.150000,11.022500


In [49]:

dest_mean_dt = dest_port_df.groupby(
    pd.to_datetime(dest_port_df['기준일자(DATE)']).dt.to_period('M')
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

dest_mean_dt

,항만효율성(PORT_EFFICIENCY),총항해시간(TOTAL_SAILING_TIME),대기시간(WAITING_TIME),항만정시성(ON_TIME_PERFORMANCE)
기준일자(DATE),,,,
2025-05,18.211111,341.750000,0.123333,3.035556
2025-06,20.884103,478.216154,0.085897,12.598718
2025-07,21.425625,440.201875,0.134687,0.596875
2025-08,20.710286,533.824000,0.165143,3.995714
2025-09,18.349231,493.987692,0.116538,6.944615
2025-10,18.570000,348.740488,0.180244,7.050000
2025-11,19.877200,461.694400,0.177600,1.176400
2025-12,19.583846,468.152564,0.074872,13.669744
2026-01,20.900667,430.941000,0.277667,8.535667


In [50]:
kpli_mean_day_dt = kpli_port_df.groupby(
    pd.to_datetime(kpli_port_df['기준일자(DATE)'])
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

# 종합의 일 평균
kpli_mean_day_dt = kpli_mean_day_dt.reset_index()


In [51]:
dest_mean_day_dt = dest_port_df.groupby(
    pd.to_datetime(dest_port_df['기준일자(DATE)'])
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

# 종합을 제외한 목정항(12개)의 일 평균
dest_mean_day_dt = dest_mean_day_dt.reset_index()

In [52]:
kpli_mean_mon_dt = kpli_mean_day_dt.groupby(
    pd.to_datetime(kpli_mean_day_dt['기준일자(DATE)']).dt.to_period('M')
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

kpli_mean_mon_dt

,항만효율성(PORT_EFFICIENCY),총항해시간(TOTAL_SAILING_TIME),대기시간(WAITING_TIME),항만정시성(ON_TIME_PERFORMANCE)
기준일자(DATE),,,,
2025-05,18.990000,239.370000,0.130000,2.290000
2025-06,20.554167,338.818333,0.076667,13.346667
2025-07,19.753333,297.745000,0.113333,3.466667
2025-08,18.922500,336.622500,0.145000,6.217500
2025-09,18.047500,318.112500,0.145000,9.135000
2025-10,18.553333,303.633333,0.162500,10.323333
2025-11,18.075000,350.335000,0.125000,7.350000
2025-12,18.690000,326.548000,0.080000,14.276000
2026-01,19.317500,334.735000,0.150000,11.022500


In [53]:

dest_mean_mon_dt = dest_mean_day_dt.groupby(
    pd.to_datetime(dest_mean_day_dt['기준일자(DATE)']).dt.to_period('M')
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

dest_mean_mon_dt

,항만효율성(PORT_EFFICIENCY),총항해시간(TOTAL_SAILING_TIME),대기시간(WAITING_TIME),항만정시성(ON_TIME_PERFORMANCE)
기준일자(DATE),,,,
2025-05,18.211111,341.750000,0.123333,3.035556
2025-06,20.672546,458.231111,0.069213,12.828704
2025-07,21.098389,419.920667,0.138167,-1.259153
2025-08,20.645868,529.333229,0.164028,4.031493
2025-09,18.565621,505.736732,0.114216,7.808889
2025-10,19.041678,338.849117,0.210847,6.580049
2025-11,19.828333,467.999833,0.162833,1.681667
2025-12,19.622262,466.461310,0.074802,13.599079
2026-01,20.396508,420.298135,0.267659,8.589817


평균값이 다르다.


최초 평균값은 일자 삭제 후 일괄 평균을 냈는데 ->
해당 데이터는 월별로 가지고 있는 일자가 다름.


때문에 일별 평균을 구한 이후 월별 평균을 구함. ->
그래도 달랐음.


때문에 더 세분화할 필요성을 느낌.  ->
따라서 목적항별 일자별 월별 로 구분하여 평균을 구함.


평균을 구하는 기준을 잘못 잡은게 아닌가



In [66]:
kpli_mean_dp_day_dt = kpli_port_df.groupby([
    '도착항코드(DEST_PORT_CODE)',
    pd.to_datetime(kpli_port_df['기준일자(DATE)'])
])[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

kpli_mean_dp_day_dt = kpli_mean_dp_day_dt.reset_index()
kpli_mean_dp_day_dt


kpli_mean_dp_mon_dt = kpli_mean_dp_day_dt.groupby([
    '도착항코드(DEST_PORT_CODE)',
    pd.to_datetime(kpli_mean_dp_day_dt['기준일자(DATE)']).dt.to_period('M')
])[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

kpli_mean_dp_mon_dt = kpli_mean_dp_mon_dt.reset_index()
kpli_mean_dp_mon_dt

kpli_mean_dmon_dt = kpli_mean_dp_mon_dt.groupby(['기준일자(DATE)']
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

kpli_mean_dmon_dt = kpli_mean_dmon_dt.reset_index()
kpli_mean_dmon_dt




,기준일자(DATE),항만효율성(PORT_EFFICIENCY),총항해시간(TOTAL_SAILING_TIME),대기시간(WAITING_TIME),항만정시성(ON_TIME_PERFORMANCE)
0,2025-05,18.990000,239.370000,0.130000,2.290000
1,2025-06,20.554167,338.818333,0.076667,13.346667
2,2025-07,19.753333,297.745000,0.113333,3.466667
3,2025-08,18.922500,336.622500,0.145000,6.217500
4,2025-09,18.047500,318.112500,0.145000,9.135000
5,2025-10,18.553333,303.633333,0.162500,10.323333
6,2025-11,18.075000,350.335000,0.125000,7.350000
7,2025-12,18.690000,326.548000,0.080000,14.276000
8,2026-01,19.317500,334.735000,0.150000,11.022500
9,2026-02,19.250000,371.660000,0.090000,4.355000


In [67]:

dest_mean_dp_day_dt = dest_port_df.groupby([
    '도착항코드(DEST_PORT_CODE)',
    pd.to_datetime(dest_port_df['기준일자(DATE)'])
])[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

# 종합을 제외한 목정항(12개)의 일 평균
dest_mean_dp_day_dt = dest_mean_dp_day_dt.reset_index()
dest_mean_dp_day_dt




dest_mean_dp_day_dt = dest_port_df.groupby([
    '도착항코드(DEST_PORT_CODE)',
    pd.to_datetime(dest_port_df['기준일자(DATE)'])
])[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

dest_mean_dp_day_dt = dest_mean_dp_day_dt.reset_index()
dest_mean_dp_day_dt


dest_mean_dp_mon_dt = dest_mean_dp_day_dt.groupby([
    '도착항코드(DEST_PORT_CODE)',
    pd.to_datetime(dest_mean_dp_day_dt['기준일자(DATE)']).dt.to_period('M')
])[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

dest_mean_dp_mon_dt = dest_mean_dp_mon_dt.reset_index()
dest_mean_dp_mon_dt

dest_mean_dmon_dt = dest_mean_dp_mon_dt.groupby(['기준일자(DATE)']
)[['항만효율성(PORT_EFFICIENCY)',
 '총항해시간(TOTAL_SAILING_TIME)',
 '대기시간(WAITING_TIME)',
 '항만정시성(ON_TIME_PERFORMANCE)']].mean()

dest_mean_dmon_dt = dest_mean_dmon_dt.reset_index()
dest_mean_dmon_dt


,기준일자(DATE),항만효율성(PORT_EFFICIENCY),총항해시간(TOTAL_SAILING_TIME),대기시간(WAITING_TIME),항만정시성(ON_TIME_PERFORMANCE)
0,2025-05,18.211111,341.750000,0.123333,3.035556
1,2025-06,21.064931,547.361944,0.069653,15.422639
2,2025-07,23.260758,587.563712,0.153939,-2.368939
3,2025-08,20.762424,565.143712,0.159015,3.500606
4,2025-09,18.283250,501.970250,0.099750,7.484250
5,2025-10,19.960741,429.149136,0.297670,4.344645
6,2025-11,20.634000,515.667750,0.143250,1.630000
7,2025-12,20.601955,604.373182,0.060409,13.929136
8,2026-01,22.385500,515.946833,0.379083,10.154583
9,2026-02,21.616061,620.815455,0.072727,1.589545



도착항코드, 일, 월 순으로 평균을 구했으나 기준일자 25년 6월 항만효율성을 기준으로 보았을때



---
| 구분 | 종합 | 개별 |
| --- | --- | --- |
|전체평균|20.472|20.884|
|일->월평균|20.554|20.672|
|도착항->일->월평균|20.554|20.064|
---
로 종합의 데이터는 사용하지 않는것이 더 맞을 것으로 판단됨.